presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1143,22 +1134,23 @@ protected Scope visitTable(Table table, Optional<Scope> scope)
 
             QualifiedObjectName name = createQualifiedObjectName(session, table, table.getName());
             analysis.addEmptyColumnReferencesForTable(accessControl, session.getIdentity(), name);
+            Optional<TableHandle> tableHandle = metadata.getTableHandle(session, name);
 
             // If this is a materialized view, get the name of the storage table
-            Optional<QualifiedName> storageName = getMaterializedViewStorageTableName(table, name);
+            Optional<QualifiedName> storageName = getMaterializedViewStorageTableName(name);
             Optional<TableHandle> storageHandle = Optional.empty();
             if (storageName.isPresent()) {
-                storageHandle = getMaterializedViewStorageTableHandle(table, storageName.get());
+                storageHandle = metadata.getTableHandle(session, createQualifiedObjectName(session, table, storageName.get()));
             }
 
             // If materialized view is current, answer the query using the storage table
             Identifier catalogName = new Identifier(name.getCatalogName());
             Identifier schemaName = new Identifier(name.getSchemaName());
             Identifier tableName = new Identifier(name.getObjectName());
-            QualifiedName mvName = QualifiedName.of(ImmutableList.of(catalogName, schemaName, tableName));
-            Optional<TableHandle> mvHandle = metadata.getTableHandle(session, createQualifiedObjectName(session, table, mvName));
-            if (storageHandle.isPresent() && metadata.isMaterializedViewCurrent(session, mvHandle.get()).getKey() == true) {
-                name = createQualifiedObjectName(session, table, storageName.get());
+            QualifiedName materializedViewName = QualifiedName.of(ImmutableList.of(catalogName, schemaName, tableName));
+            Optional<TableHandle> materializedViewHandle = metadata.getTableHandle(session, createQualifiedObjectName(session, table, materializedViewName));
+            if (storageHandle.isPresent() && metadata.isMaterializedViewCurrent(session, materializedViewHandle.get()).getIsFresh()) {
+                tableHandle = storageHandle;
             }
             else {
                 // Either this is a reference to a logical view or the materialized view is not current and should be expanded like a logical view
```

Comment:
@martint I wanted to point out that, here we read materialized view like a logical view since it is stale. Another option would be: 
Analyze materialized view to get columns and other attributes and create a `new` logical view and create scope for that. This analysis is necessary since the `columns` exposed by the materialized view is a single dummy column of type `varchar`. (This is similar to how a presto view exposes a single dummy column of type varchar).
Storing actual column list at create materialized view time also poses issues since the column list needs to be a list of hive metastore columns and the connector does not have the correct TypeManager to do that conversion from SPI TypeId to hive column type.

Reply:
After our offline discussion. We analyze the materialized view and create scope for it similar to a logical view. This also checks for recursive materialized view definitions etc.